src/MicrosoftAuthFlow.js

```diff
@@ -0,0 +1,172 @@
+const fs = require('fs')
+const path = require('path')
+const crypto = require('crypto')
+
+const minecraftFolderPath = require('minecraft-folder-path')
+const debug = require('debug')('xboxlive-auth')
+
+const { Authentication, msalConfig } = require('./common/Constants')
+
+const LiveTokenManager = require('./TokenManagers/LiveTokenManager')
+const JavaTokenManager = require('./TokenManagers/JavaTokenManager')
+const XboxTokenManager = require('./TokenManagers/XboxTokenManager')
+const MsaTokenManager = require('./TokenManagers/MsaTokenManager')
+const BedrockTokenManager = require('./TokenManagers/BedrockTokenManager')
+
+async function retry (methodFn, beforeRetry, times) {
+  while (times--) {
+    if (times !== 0) {
+      try { return await methodFn() } catch (e) { debug(e) }
+      await new Promise(resolve => setTimeout(resolve, 2000))
+      await beforeRetry()
+    } else {
+      return await methodFn()
+    }
+  }
+}
+
+class MicrosoftAuthFlow {
+  constructor (username, cacheDir = minecraftFolderPath, options = {}, codeCallback) {
+    this.username = username
+    this.options = options
+    this.initTokenCaches(username, cacheDir)
+    this.codeCallback = codeCallback
+  }
+
+  initTokenCaches (username, cacheDir) {
+    const hash = sha1(username).substr(0, 6)
+
+    let cache = path.join(cacheDir, 'npm-cache')
+    debug(`Using cache path: ${cache}`)
+    try {
+      if (!fs.existsSync(cache)) {
+        fs.mkdirSync(cache, { recursive: true })
+      }
+    } catch (e) {
+      console.log('Failed to open cache dir', e)
+      cache = __dirname
+    }
+
+    const cachePaths = {
+      live: path.join(cache, `./${hash}_live-cache.json`),
+      msa: path.join(cache, `./${hash}_msa-cache.json`),
+      xbl: path.join(cache, `./${hash}_xbl-cache.json`),
+      mca: path.join(cache, `./${hash}_mca-cache.json`),
+      mba: path.join(cache, `./${hash}_mca-cache.json`)
+    }
+
+    if (this.options.authTitle) { // Login with login.live.com
+      const scopes = ['service::user.auth.xboxlive.com::MBI_SSL']
+      this.msa = new LiveTokenManager(this.options.authTitle, scopes, cachePaths.live)
+    } else { // Login with microsoftonline.com
+      const scopes = ['XboxLive.signin', 'offline_access']
+      this.msa = new MsaTokenManager(msalConfig, scopes, cachePaths.msa)
+    }
+
+    const keyPair = crypto.generateKeyPairSync('ec', { namedCurve: 'P-256' })
+    this.xbl = new XboxTokenManager(Authentication.XSTSRelyingParty, keyPair, cachePaths.xbl)
+
+    if (this.options.authTitle) { // Login with bedrock
```

Comment:
Auth title is used for both Microsoft and Xbox live auth steps. It’s not related to any version of Minecraft. It’s the client ID in the ms auth step, which would probably be a better name. This is always set, this is just an option to customize it since on Bedrock edition, servers get access to extra auth data from the Xbox step which they can choose to verify although vanilla doesn’t. What I was mentioning is keeping a list of known client IDs, but it’s probably not worth it.

Reply:
How should we identify which one we want?